src/run.rs

```diff
@@ -0,0 +1,202 @@
+
+    let handle = thread::spawn(move || {
+        let mut handles = Vec::new();
+        let mut statuses = Vec::new();
+        for cmd in commands {
+            handles.push(run_command(
+                cmd,
+                send.clone(),
+                options.clone(),
+                kill_trigger_clone.clone(),
+            ));
+        }
+
+        for (idx, handle) in handles.into_iter().enumerate() {
+            statuses.push(handle.join().unwrap_or((command_names[idx].clone(), None)));
+        }
+
+        statuses
+    });
+
+    CommandHandle {
+        handle,
+        channel: recv,
+        kill_trigger,
+    }
+}
+
+fn run_command(
+    command: InnerCommand,
+    send_chan: mpsc::Sender<OutputMessage>,
+    options: Options,
+    kill_trigger: kill_barrier::KillBarrier,
+) -> thread::JoinHandle<ExitResult> {
+    thread::spawn(move || loop {
+        let mut command_process = command.to_stdlib_command();
+        let command_name = command.name.clone();
+
+        let _ = send_chan.send(OutputMessage {
+            name: command_name.clone(),
+            message: OutputMessagePayload::Start,
+        });
+
+        let cmd_handle = command_process.spawn();
+        if cmd_handle.is_err() {
+            let _ = send_chan.send(OutputMessage {
+                name: command_name.clone(),
+                message: OutputMessagePayload::Error(cmd_handle.err().unwrap()),
+            });
+
+            match options.restart {
+                RestartOptions::Restart => continue,
+                RestartOptions::Kill => {
+                    let _ = kill_trigger.initiate_kill();
+                    return (command_name, None);
+                }
+                RestartOptions::Continue => return (command_name, None),
+            }
+        }
+
+        let mut cmd_handle = cmd_handle.unwrap();
+
+        let std_out = cmd_handle.stdout.take();
+        let std_err = cmd_handle.stderr.take();
+        let mut std_out_handle = None;
+        let mut std_err_handle = None;
+
+        let shared_handle = Arc::new(Mutex::new(cmd_handle));
+
+        let child_clone = shared_handle.clone();
+        let kill_trigger_clone = kill_trigger.clone();
+        thread::spawn(move || kill_thread(&kill_trigger_clone, child_clone));
+
+        if let Some(output) = std_out {
+            let mut buffered_stdout = BufReader::new(output);
+            let new_name = command_name.clone();
+            let new_chan = send_chan.clone();
+            std_out_handle = Some(thread::spawn(move || {
+                read_stream(&new_name, new_chan, &mut buffered_stdout, true);
+            }));
+        }
+
+        if let Some(output) = std_err {
+            let mut buffered_stdout = BufReader::new(output);
+            let new_name = command_name.clone();
+            let new_chan = send_chan.clone();
+            std_err_handle = Some(thread::spawn(move || {
+                read_stream(&new_name, new_chan, &mut buffered_stdout, false);
```

Comment:
Why not consume `buffered_stdout` here?

Reply:
It can, I can change it. I usually prefer to not consume if possible. It leads to a more flexible API that can be slightly faster in certain cases

Reply:
I see. You don't want to make `read_stream` unnecessarily consume the variable, even though in this case the calling code has no use for the variable anymore and could happily give up ownership. That makes sense. I agree that `read_stream` should avoid being too "greedy" here.